File: src/infrastructure/tools/search/search-text-in-files.ts
/**
 * search_text_in_files — 按内容搜索（支持正则，自动跳过二进制文件并遵循 .gitignore）
 *
 * 底层使用 Rust 的 ripgrep 核心库，通过 Tauri invoke 调用。
 */
import { invoke } from '@tauri-apps/api/core'
import { withCancelResult } from '@/utils/withCancel'
import { toolRegistry } from '@/domain/tools'
import type { ToolExecutor, ToolContext, ToolResult } from '@/domain/tools/types'
import { t } from '@/ui/i18n'
import { createSearchTask, resolveSearchRoot } from './common'

interface TextSearchResult {
  path: string
  line_number: number
  line: string
}

toolRegistry.register(
    'search_text_in_files',
    (async (args: Record<string, any>, ctx: ToolContext): Promise<ToolResult> => {
    // 参数校验：query 为必填
    if (!args.query) {
      return {
        content:
          'Missing required parameter: "query". Please provide a text or regex pattern to search for.',
        uiData: { length: 0 },
      }
    }

    const { taskId, stop } = createSearchTask('search_text', ctx.abortSignal)
    let cancelled = false

    const results = await withCancelResult(
      ctx.abortSignal,
      invoke<TextSearchResult[]>('search_text_in_files', {
        root: await resolveSearchRoot(args.path, ctx.sessionId),
        query: args.query,
        maxResults: args.max_results ?? 30,
        taskId,
      }),
      () => {
        cancelled = true
        stop()
        return [] as TextSearchResult[]
      },
    )

    if (cancelled) {
      throw `[Search cancelled] Search for "${args.query}" was cancelled.`
    }

    // 防御：invoke 可能返回 null/undefined，统一转为数组
    const safeResults: TextSearchResult[] = results ?? []

    const MAX_CHARS = 32000
    let output = `🔍 ${safeResults.length} match(es) for "${args.query}":\n`

    for (const r of safeResults) {
      const line = `  📄 ${r.path}:${r.line_number}  ${r.line.trim()}`
      if (output.length + line.length + 1 > MAX_CHARS) {
        output += `\n... (truncated, ${safeResults.length} total matches)`
        break
      }
      output += line + '\n'
    }

    return {
      content: output,
      uiData: {
        length: safeResults.length,
      },
    }
  }) as ToolExecutor,
    t('搜索关键字'),
)

File: src-tauri/src/agent/native_tools/search/search_text_in_files.rs
//! `search_text_in_files` 工具（原生）— 按文件内容（正则）搜索，输出带 32000 字符截断。

use crate::agent::native_tools::common::{arg_i64, arg_str, resolve_safe_path};
use crate::agent::native_tools::{NativeToolCtx, NativeToolOutcome};
use serde_json::{json, Value};
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;

pub(crate) async fn search_text_in_files_tool(
    ctx: &NativeToolCtx<'_>,
    args: &Value,
) -> Result<NativeToolOutcome, String> {
    let query = arg_str(args, "query").unwrap_or_default();
    if query.trim().is_empty() {
        return Ok(NativeToolOutcome::Value {
            content: "Missing required parameter: \"query\". Please provide a text or regex pattern to search for.".to_string(),
            ui_data: Some(json!({ "length": 0 })),
        });
    }

    let root = resolve_safe_path(&arg_str(args, "path").unwrap_or_else(|| ".".into()), "r", ctx.security)?;
    let max_results = arg_i64(args, "max_results").unwrap_or(30).clamp(1, 500) as usize;

    let cancel_flag = Arc::new(AtomicBool::new(false));
    let flag = cancel_flag.clone();
    let root_c = root.clone();
    let q = query.clone();
    let task = tokio::task::spawn_blocking(move || {
        crate::search::search_text_in_files(&root_c, &q, max_results, &flag)
    });

    let results = tokio::select! {
        _ = ctx.cancel.cancelled() => {
            cancel_flag.store(true, Ordering::SeqCst);
            return Ok(NativeToolOutcome::error(format!("[Search cancelled] Search for \"{}\" was cancelled.", query)));
        }
        r = task => r.map_err(|e| format!("Search failed: {}", e))?,
    };

    const MAX_CHARS: usize = 32000;
    let mut output = format!("🔍 {} match(es) for \"{}\":\n", results.len(), query);
    for r in &results {
        let line = format!("  📄 {}:{}  {}", r.path, r.line_number, r.line.trim());
        if output.len() + line.len() + 1 > MAX_CHARS {
            output.push_str(&format!("\n... (truncated, {} total matches)", results.len()));
            break;
        }
        output.push_str(&line);
        output.push('\n');
    }

    Ok(NativeToolOutcome::Value {
        content: output,
        ui_data: Some(json!({ "length": results.len() })),
    })
}
